File: ige-frontend/src/messages.rs
// We should have some way of threading events and controlmsgs
// so that we can have responses to specific messages

/// The type of messages we send from Haskell to Rust
#[derive(Serialize, Deserialize, Debug, Copy, Clone)]
pub enum ControlMsg {
    DispMsg (DisplayCommand),
    GraphMsg (GraphCommand)
}

#[derive(Serialize, Deserialize, Debug, Copy, Clone)]
pub enum GraphCommand {
    AddNode,
    DelNode {
        node_id: u32
    },
    AddEdge {
        node_id1: u32,
        node_id2: u32
    },
    DelEdge {
        edge_id: u32,
    }
}

#[derive(Serialize, Deserialize, Debug, Copy, Clone)]
pub enum DisplayCommand {
    Refresh,
    Zoom { percent: f64 },
    Rotate { radians: f64 },
    Translate {
        x: f64,
        y: f64
    },
    GetSelection
}

/// The type of messages we send from Rust to Haskell
#[derive(Serialize, Deserialize, Debug, Copy, Clone)]
pub enum Event {
    KeyPress {
        key: u32,
        modifier: u32
    },
    Selection {
        node_id: u32
    }
}

File: ige-backend/Messages.hs
{-# LANGUAGE DeriveGeneric #-}

module Messages
(
  ControlMsg(..),
  DisplayCommand(..),
  GraphCommand(..),
  Event(..)
)
where

import Data.MessagePack
import GHC.Generics
import Data.Binary


data ControlMsg =
  DispMsg DisplayCommand
  | GraphMsg GraphCommand
  deriving (Show, Eq, Generic)

instance MessagePack ControlMsg where

instance Binary ControlMsg where
  get = get >>= fromObject
  put = put . toObject

data GraphCommand =
  AddNode
  | DelNode { nodeId :: Int }
  | AddEdge { nodeId1 :: Int, nodeId2 :: Int }
  | DelEdge { edgeId :: Int }
  deriving (Show, Eq, Generic)

instance MessagePack GraphCommand where

instance Binary GraphCommand where
  get = get >>= fromObject
  put = put . toObject

data DisplayCommand =
  Refresh
  | Zoom { percent :: Double }
  | Rotate { radians :: Double }
  | Translate { x :: Double, y :: Double }
  | GetSelection
  deriving (Show, Eq, Generic)

instance MessagePack DisplayCommand where

instance Binary DisplayCommand where
  get = get >>= fromObject
  put = put . toObject

data Event =
  KeyPress { key :: Int, modifier :: Int }
  | Selection { selNodeId :: Int }
  deriving (Show, Eq, Generic)

instance MessagePack Event where

instance Binary Event where
  get = get >>= fromObject
  put = put . toObject
